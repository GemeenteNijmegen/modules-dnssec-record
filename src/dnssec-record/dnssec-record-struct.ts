import {
  Duration,
  CustomResource,
  aws_route53 as route53,
  aws_iam as iam,
  custom_resources as customresource,
  aws_logs as logs,
} from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { DnssecRecordFunction } from './dnssec-record-function';


export interface DnssecRecordStructProps {
  readonly keySigningKey: route53.CfnKeySigningKey;
  readonly hostedZone: route53.IHostedZone;
  readonly parentHostedZone: route53.IHostedZone;
}

export class DnssecRecordStruct extends Construct {


  constructor(scope: Construct, id: string, props: DnssecRecordStructProps) {
    super(scope, id);

    const lambda = new DnssecRecordFunction(this, 'lambda', {
      timeout: Duration.seconds(60),
    });

    // Allow to request the state of a change in route53
    lambda.addToRolePolicy(new iam.PolicyStatement({
      effect: iam.Effect.ALLOW,
      actions: [
        'route53:GetChange',
      ],
      resources: ['*'],
    }));

    // Allow obtaining info of DNSSEC on hosted zone
    lambda.addToRolePolicy(new iam.PolicyStatement({
      effect: iam.Effect.ALLOW,
      actions: [
        'route53:GetDNSSEC',
      ],
      resources: [
        props.hostedZone.hostedZoneArn,
      ],
    }));

    // Allow UPSERTING records on parent hosted zone
    lambda.addToRolePolicy(new iam.PolicyStatement({
      effect: iam.Effect.ALLOW,
      actions: [
        'route53:ChangeResourceRecordSets',
      ],
      resources: [
        props.parentHostedZone.hostedZoneArn,
      ],
    }));

    const customResourceProvider = new customresource.Provider(this, 'provider', {
      onEventHandler: lambda,
      logRetention: logs.RetentionDays.ONE_MONTH,
    });

    new CustomResource(this, 'custom-resource', {
      serviceToken: customResourceProvider.serviceToken,
      properties: {
        keySigningKeyName: props.keySigningKey.name,
        hostedZoneId: props.hostedZone.hostedZoneId,
        hostedZoneName: props.hostedZone.zoneName,
        parentHostedZoneId: props.parentHostedZone.hostedZoneId,
      },
    });

  }


}