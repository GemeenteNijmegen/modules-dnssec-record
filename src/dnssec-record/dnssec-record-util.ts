
import {
  Route53Client,
  GetDNSSECCommand,
  ChangeResourceRecordSetsCommand,
  GetChangeCommand,
  ChangeStatus,
  KeySigningKey,
} from '@aws-sdk/client-route-53';

export class DnssecRecordUtil {

  private route53Client: Route53Client;

  constructor(client: Route53Client) {
    this.route53Client = client;
  }

  /**
   * Will search the hosted zone for active KeySigningKeys and
   * returns the DS record value from the first active KSK found.
   * @param hostedZoneId
   * @returns
   */
  async getDsRecordValue(hostedZoneId: string, keySigningKeyName: string) {
    // Obtain DNSSEC status from hosted zone
    const dnssecCommand = new GetDNSSECCommand({
      HostedZoneId: hostedZoneId,
    });
    const dnssecInfo = await this.route53Client.send(dnssecCommand);

    // Find the key signing key by name
    const keySigningKey = dnssecInfo.KeySigningKeys?.find((k: KeySigningKey) => k.Name == keySigningKeyName && k.Status == 'ACTIVE');
    if (!keySigningKey) {
      console.error('Could not find active KSK', keySigningKeyName, JSON.stringify(dnssecInfo.KeySigningKeys));
      throw Error(`Could not find an (active) KeySignigKey with name ${keySigningKeyName} (see logging for more details)`);
    }

    // Get the record value
    const dsRecordValue = keySigningKey.DSRecord;
    if (!dsRecordValue) {
      throw Error(`Value: ${dsRecordValue} is not a valid DS record value`);
    }

    return dsRecordValue;
  }


  /**
   * Tries to create the DS record in the parent hosted zone
   * @param parentHostedZoneId
   * @param dsRecordName
   * @param dsRecordValue
   */
  async createDsRecord(parentHostedZoneId: string, dsRecordName: string, dsRecordValue: string) {
    const createRecordCommand = new ChangeResourceRecordSetsCommand({
      HostedZoneId: parentHostedZoneId, // DS is in parent hosted zone!
      ChangeBatch: {
        Comment: 'Create DS record in parent hosted zone',
        Changes: [
          {
            Action: 'UPSERT', // Create if not exist update otherwise
            ResourceRecordSet: {
              Name: dsRecordName,
              Type: 'DS',
              ResourceRecords: [
                { Value: dsRecordValue },
              ],
              TTL: 3600,
            },
          },
        ],
      },
    });
    const createRecordResult = await this.route53Client.send(createRecordCommand);

    const changeId = createRecordResult.ChangeInfo?.Id;
    if (!changeId) {
      throw Error('No change ID returned while creating record');
    }

    return changeId;
  }


  /**
   * Wait for a chagne to be INSYNC
   * @param changeId
   * @param maxRetries
   * @param intervalMilis
   * @returns
   */
  async waitForChange(changeId: string, maxRetries: number, intervalMilis: number) {

    let retries = 0;
    while (retries <= maxRetries) {
      retries++;

      // Check if change is INSYNC
      const changeStatusCommand = new GetChangeCommand({
        Id: changeId,
      });
      const status = await this.route53Client.send(changeStatusCommand);

      if (status.ChangeInfo?.Status == ChangeStatus.INSYNC) {
        return true;
      }

      await this.sleep(intervalMilis);
    }
    return false;
  }

  private async sleep(millis: number) {
    return new Promise(resolve => setTimeout(resolve, millis));
  }

}