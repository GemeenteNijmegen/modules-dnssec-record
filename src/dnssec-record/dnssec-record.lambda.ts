import { Route53Client } from '@aws-sdk/client-route-53';
import {
  CloudFormationCustomResourceEvent,
  CloudFormationCustomResourceResponse,
} from 'aws-lambda';
import { DnssecRecordUtil } from './dnssec-record-util';

const route53Client = new Route53Client({});

/**
 * Entry point of the custom resource
 */
exports.handler = async (event: CloudFormationCustomResourceEvent): Promise<CloudFormationCustomResourceResponse> => {
  switch (event.RequestType) {
    case 'Create': return onCreateUpdate(event);
    case 'Update': return onCreateUpdate(event);
    case 'Delete': return response('SUCCESS', event, 'Note: DS record is not deleted from hosted zone');
  }
};

async function onCreateUpdate(
  event: CloudFormationCustomResourceEvent,
): Promise<CloudFormationCustomResourceResponse> {

  // Get parameters for custom resource
  const keySigningKeyName = event.ResourceProperties.keySigningKeyName;
  const hostedZoneId = event.ResourceProperties.hostedZoneId;
  const hostedZoneName = event.ResourceProperties.hostedZoneName;
  const parentHostedZoneId = event.ResourceProperties.parentHostedZoneId;

  console.info('KeySigningKey name:', keySigningKeyName);
  console.info('Hosted zone ID:', hostedZoneId);
  console.info('Hosted zone name:', hostedZoneName);
  console.info('Parent hosted zone ID:', parentHostedZoneId);

  const util = new DnssecRecordUtil(route53Client);

  try {

    const dsRecordValue = await util.getDsRecordValue(hostedZoneId, keySigningKeyName);
    const changeId = await util.createDsRecord(parentHostedZoneId, hostedZoneName, dsRecordValue);
    const successful = await util.waitForChange(changeId, 2, 3000);

    if (successful) {
      return response('SUCCESS', event);
    }

    return response('SUCCESS', event, 'Added DS record however change stats is not yet INSYNC');

  } catch (error) {
    console.error(error);
    return response('FAILED', event, JSON.stringify(error));
  }

}


/**
 * Return reponse to CloudFormation
 * @param status
 * @param event
 * @param reason
 * @returns
 */
function response(
  status: 'SUCCESS' | 'FAILED',
  event: CloudFormationCustomResourceEvent,
  reason?: string,
): CloudFormationCustomResourceResponse {
  return {
    Status: status,
    Reason: reason ? reason : '',
    LogicalResourceId: event.LogicalResourceId,
    PhysicalResourceId: event.ResourceProperties.PhysicalResourceId,
    RequestId: event.RequestId,
    StackId: event.StackId,
  };
}

