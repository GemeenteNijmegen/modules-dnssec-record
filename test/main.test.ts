import {
  App,
  Stack,
  aws_route53 as route53,
} from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { Construct } from 'constructs';
import { DnssecRecordStruct } from '../src/dnssec-record/dnssec-record-struct';

test('snapshot', () => {

  class MyStack extends Stack {
    constructor(scope: Construct, id: string) {
      super(scope, id);

      const rootHz = route53.HostedZone.fromHostedZoneAttributes(this, 'root-hz', {
        hostedZoneId: 'OEIJGEOIJG2380JEFO',
        zoneName: 'example.com',
      });

      const subHz = route53.HostedZone.fromHostedZoneAttributes(this, 'sub-hz', {
        hostedZoneId: 'OEIJGEOIJG2380JEF1',
        zoneName: 'sub.example.com',
      });

      const ksk = new route53.CfnKeySigningKey(this, 'dnssec-keysigning-key', {
        name: 'ksk_name',
        status: 'ACTIVE',
        hostedZoneId: 'OEIJGEOIJG2380JEFO',
        keyManagementServiceArn: 'arn:aws:some:kms/key',
      });

      new DnssecRecordStruct(this, 'record', {
        keySigningKey: ksk,
        hostedZone: subHz,
        parentHostedZone: rootHz,
      });
    }
  }

  const app = new App();
  const stack = new MyStack(app, 'stack');

  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();

});