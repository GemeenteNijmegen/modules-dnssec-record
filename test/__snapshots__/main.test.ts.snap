// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": Object {
      "DependsOn": Array [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "b120b13d9d868c7622e7db1b68bae4c0f82ffd0227b8c15f2cef38e186ff3827.zip",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": Array [
          Object {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "dnsseckeysigningkey": Object {
      "Properties": Object {
        "HostedZoneId": "OEIJGEOIJG2380JEFO",
        "KeyManagementServiceArn": "arn:aws:some:kms/key",
        "Name": "ksk_name",
        "Status": "ACTIVE",
      },
      "Type": "AWS::Route53::KeySigningKey",
    },
    "recordcustomresourceEA77BEBE": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "recordproviderframeworkonEvent357A600A",
            "Arn",
          ],
        },
        "hostedZoneId": "OEIJGEOIJG2380JEF1",
        "hostedZoneName": "sub.example.com",
        "keySigningKeyName": "ksk_name",
        "parentHostedZoneId": "OEIJGEOIJG2380JEFO",
      },
      "Type": "AWS::CloudFormation::CustomResource",
      "UpdateReplacePolicy": "Delete",
    },
    "recordlambdaEA9BD739": Object {
      "DependsOn": Array [
        "recordlambdaServiceRoleDefaultPolicy3EABB022",
        "recordlambdaServiceRole7D7B7AFC",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "fa8e23e82e9434df251c2ea22a0d37b2ef9554dc226369f864b76ae41545ea6b.zip",
        },
        "Description": "src/dnssec-record/dnssec-record.lambda.ts",
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "recordlambdaServiceRole7D7B7AFC",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "recordlambdaServiceRole7D7B7AFC": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "recordlambdaServiceRoleDefaultPolicy3EABB022": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "route53:GetChange",
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "route53:GetDNSSEC",
                "route53:ChangeResourceRecordSets",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":route53:::hostedzone/OEIJGEOIJG2380JEF1",
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":route53:::hostedzone/OEIJGEOIJG2380JEFO",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "recordlambdaServiceRoleDefaultPolicy3EABB022",
        "Roles": Array [
          Object {
            "Ref": "recordlambdaServiceRole7D7B7AFC",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "recordproviderframeworkonEvent357A600A": Object {
      "DependsOn": Array [
        "recordproviderframeworkonEventServiceRoleDefaultPolicyEDF16A12",
        "recordproviderframeworkonEventServiceRole4FF80382",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "c691172cdeefa2c91b5a2907f9d81118e47597634943344795f1a844192dd49c.zip",
        },
        "Description": "AWS CDK resource provider framework - onEvent (stack/record/provider)",
        "Environment": Object {
          "Variables": Object {
            "USER_ON_EVENT_FUNCTION_ARN": Object {
              "Fn::GetAtt": Array [
                "recordlambdaEA9BD739",
                "Arn",
              ],
            },
          },
        },
        "Handler": "framework.onEvent",
        "Role": Object {
          "Fn::GetAtt": Array [
            "recordproviderframeworkonEventServiceRole4FF80382",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "recordproviderframeworkonEventLogRetention9CC1BE8E": Object {
      "Properties": Object {
        "LogGroupName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "/aws/lambda/",
              Object {
                "Ref": "recordproviderframeworkonEvent357A600A",
              },
            ],
          ],
        },
        "RetentionInDays": 30,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "recordproviderframeworkonEventServiceRole4FF80382": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "recordproviderframeworkonEventServiceRoleDefaultPolicyEDF16A12": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "recordlambdaEA9BD739",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "recordproviderframeworkonEventServiceRoleDefaultPolicyEDF16A12",
        "Roles": Array [
          Object {
            "Ref": "recordproviderframeworkonEventServiceRole4FF80382",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
